### PR TITLE
Adds `timeout_ms` and a new method `MultiGet` to InternalKVAccessor.

### PR DESCRIPTION
**Pull request by rynewang**
This is to ensure feature parity of the accessor vs the PythonGcsClient, as a preparation step to replace the latter. Previously it sets all timeouts to a SYS_gcs_server_request_timeout_seconds but now we make it user controllable.

Semantics of the new field: for each Async methods in `accessor.h`, the new field is `const int64_t timeout_ms`. `-1` means infinite timeout; `> 0` means the specified timeout in milliseconds.

Also adds a method MultiGet.

### CODE REVIEW

File: src/ray/gcs/gcs_client/accessor.cc [hunk 1]
```diff
@@ -21,10 +21,23 @@
...
-inline int64_t GetGcsTimeoutMs() {
+
+inline int64_t GetDefaultGcsTimeoutMs() {
   return absl::ToInt64Milliseconds(
       absl::Seconds(RayConfig::instance().gcs_server_request_timeout_seconds()));
 }
+
+/// Gets the timeout in milliseconds, used by a GCS Client RPC call.
+/// -1 means infinite retry.
+/// 0 means system default (= RAY_gcs_server_request_timeout_seconds)
```
Comment (jjyao) [inline src/ray/gcs/gcs_client/accessor.cc:32]: Should we just ask callers to pass in `GetDefaultGcsTimeoutMs()` if they want the default so we don't need to assign special meaning for `0` here.

If you check gcs_rpc_client.h, only -1 has special meaning.

**Comment by can-anyscale** [comment]
i rebased your PR to a more recent master to have better coverage in microcheck, thankks

File: src/ray/gcs/gcs_client/accessor.cc (hunk 1 above)
Comment (rynewang) [reply src/ray/gcs/gcs_client/accessor.cc]: done

File: src/ray/gcs/gcs_client/global_state_accessor.cc [hunk 2]
```diff
@@ -350,7 +350,7 @@ std::unique_ptr<std::string> GlobalStateAccessor::GetInternalKV(const std::strin
   absl::ReaderMutexLock lock(&mutex_);
   std::string value;
 
-  Status status = gcs_client_->InternalKV().Get(ns, key, value);
+  Status status = gcs_client_->InternalKV().Get(ns, key, /*timeout_ms=*/0, value);
```
Comment (jjyao) [inline src/ray/gcs/gcs_client/global_state_accessor.cc:353]: 0?

File: src/ray/gcs/gcs_client/test/usage_stats_client_test.cc [hunk 3]
```diff
@@ -88,7 +88,7 @@ TEST_F(UsageStatsClientTest, TestRecordExtraUsageTag) {
       [this]() {
         std::string value;
         RAY_CHECK_OK(this->gcs_client_->InternalKV().Get(
-            "usage_stats", "extra_usage_tag__test1", value));
+            "usage_stats", "extra_usage_tag__test1", /*timeout_ms=*/0, value));
```
Comment (jjyao) [inline src/ray/gcs/gcs_client/test/usage_stats_client_test.cc:91]: 0?

File: src/ray/gcs/gcs_client/accessor.h [hunk 4]
```diff
@@ -683,119 +687,174 @@ class InternalKVAccessor {
...
+                                    const int64_t timeout_ms,
                                     const StatusCallback &callback);
 
   // These are sync functions of the async above
 
   /// List keys with prefix stored in internal kv
   ///
-  /// The RPC will timeout after the default GCS RPC timeout is exceeded.
+  /// The RPC will timeout after the timeout_ms, or if not set, the default GCS RPC
+  /// timeout is exceeded.
```
[the remark this answers is not in the document]
Comment (rynewang) [reply src/ray/gcs/gcs_client/accessor.h:769]: updated

File: src/ray/gcs/gcs_client/accessor.cc [hunk 5]
```diff
@@ -1033,14 +1032,44 @@ Status InternalKVAccessor::AsyncInternalKVGet(
           callback(status, reply.value());
         }
       },
-      /*timeout_ms*/ GetGcsTimeoutMs());
+      timeout_ms);
+  return Status::OK();
+}
+Status InternalKVAccessor::AsyncInternalKVMultiGet(
```
[the remark this answers is not in the document]
Comment (rynewang) [reply src/ray/gcs/gcs_client/accessor.cc:1038]: done

File: src/ray/gcs/gcs_client/test/usage_stats_client_test.cc (hunk 3 above)
Comment (rynewang) [reply src/ray/gcs/gcs_client/test/usage_stats_client_test.cc:91]: fixed

**Review by jjyao — APPROVED** [review]

**Comment by rynewang** [comment]
This is ready to merge

**Pull request merged**